lib/http.c

```diff
@@ -1543,25 +1543,45 @@ static CURLcode add_haproxy_protocol_header(struct Curl_easy *data)
   DEBUGASSERT(data->conn);
   Curl_dyn_init(&req, DYN_HAXPROXY);
 
+  if(data->set.haproxyprotocol == 1) {
+    /* sending HAProxy PROXY protocol v1 header */
 #ifdef USE_UNIX_SOCKETS
-  if(data->conn->unix_domain_socket)
-    /* the buffer is large enough to hold this! */
-    result = Curl_dyn_add(&req, "PROXY UNKNOWN\r\n");
-  else {
+    if(data->conn->unix_domain_socket)
+      /* the buffer is large enough to hold this! */
+      result = Curl_dyn_add(&req, "PROXY UNKNOWN\r\n");
+    else {
 #endif
-  /* Emit the correct prefix for IPv6 */
-  tcp_version = data->conn->bits.ipv6 ? "TCP6" : "TCP4";
+    /* Emit the correct prefix for IPv6 */
+    tcp_version = data->conn->bits.ipv6 ? "TCP6" : "TCP4";
 
-  result = Curl_dyn_addf(&req, "PROXY %s %s %s %i %i\r\n",
-                         tcp_version,
-                         data->info.conn_local_ip,
-                         data->info.conn_primary_ip,
-                         data->info.conn_local_port,
-                         data->info.conn_primary_port);
+    result = Curl_dyn_addf(&req, "PROXY %s %s %s %i %i\r\n",
+                          tcp_version,
+                          data->info.conn_local_ip,
+                          data->info.conn_primary_ip,
+                          data->info.conn_local_port,
+                          data->info.conn_primary_port);
 
 #ifdef USE_UNIX_SOCKETS
-  }
+    }
 #endif
+  }
+  else {
+    /* sending HAProxy PROXY protocol v2 header */
+    result = Curl_dyn_addn(
+      &req,
+      (char []) {
+        /* signature */
+        0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A,
+        /* command is LOCAL and address family is AF_UNSPEC */
+        0x20,
+        /* protocol is UNSPEC */
+        0x00,
+        /* there is no additionnal data */
+        0x00, 0x00
+        },
```

Comment:
How about making this chunk a `const static` array instead? Will also make it slightly more readable I think.